### PR TITLE
Remove every call to BN's mutating methods

### PR DESCRIPTION
**Pull request by alcuadrado**
See #605 for its discussion.

### CODE REVIEW

**Comment by holgerd77** [comment]
@alcuadrado I've lost a bit track here, what is the state of this respectively the outcome of the discussion along? @s1na if you have got a moment your comment would also be appreciated.

**Comment by alcuadrado** [comment]
I still think that mutating BN's is risky, especially when we are using those in event handlers. 

This change is outdated, so I'd should redo it if it were to be merged. It only took me a few minutes though. Also, using these methods should result in a linting error, or they'll eventually be reintroduced.

**Comment by axic** [comment]
I'm not sure if this was ever measured, but they were added intentionally to increase performance. The non-mutating versions do an internal clone. Which is practically what you said would be harmful in #605.

Is returning a cloned stack in `step` an option?

File: lib/evm/precompiles/04-identity.ts [hunk 1]
```diff
@@ -8,9 +8,9 @@ export default function(opts: PrecompileInput): ExecResult {
 
   const data = opts.data
 
-  const gasUsed = new BN(opts._common.param('gasPrices', 'identity'))
-  gasUsed.iadd(
-    new BN(opts._common.param('gasPrices', 'identityWord')).imuln(Math.ceil(data.length / 32)),
+  let gasUsed = new BN(opts._common.param('gasPrices', 'identity'))
```
Comment (axic) [inline lib/evm/precompiles/04-identity.ts]: The precompiles don't affect stack elements, so probably no need to change them?

**Comment by alcuadrado** [comment]
> I'm not sure if this was ever measured, but they were added intentionally to increase performance.

I didn't know that. It'd be good to measure its impact before merging this change.

> Which is practically what you said would be harmful in #605.
>
> Is returning a cloned stack in `step` an option?

I think what was being proposed in #605 was cloning every element passed to the `step` event. That includes things like the stack, which consists in multiple elements, so you'd be cloning significantly more than by implementing this change. Even if you are cloning the stack, a shallow vs a deep clone can also make a big difference. I may be wrong though, predicting performance implications of systems this complex can be tricky.

> I am actually confused why would anyone think that an internal data piece returned on `step` is valid after the next call. It should be the responsibility of the caller to save that state if it wishes to reuse it.

I'm not entirely against this, but it should be documented, and it may have pretty much the same effect on the system performance as cloning things ourselves.

File: lib/evm/precompiles/04-identity.ts (hunk 1 above)
Comment (alcuadrado) [reply lib/evm/precompiles/04-identity.ts]: Yeah, I noticed the same. I changed them with the idea of also proposing a linting rule forbidding them. Without that proposal nor an explanation they must look confusing, sorry about that.

**Comment by axic** [comment]
> > I am actually confused why would anyone think that an internal data piece returned on step is valid after the next call. It should be the responsibility of the caller to save that state if it wishes to reuse it.
>
> I'm not entirely against this, but it should be documented, and it may have pretty much the same effect on the system performance as cloning things ourselves.

If it is handled by the caller, the effect is smaller. I'd imagine that in many cases the stack is displayed and not used again.

Probably the right question to ask in general: what is the more common use case of ethereumjs-vm. Is debugging on by default most of the time (remix, ganache/truffle)?

If the answer is yes, then making the impact on debugging less makes sense.

**Comment by axic** [comment]
Oh btw, the stack items were cloned outside of `runCode` after `opFns` returned, in the past. Is that not the case?

**Comment by alcuadrado** [comment]
> Probably the right question to ask in general: what is the more common use case of ethereumjs-vm. Is debugging on by default most of the time (remix, ganache/truffle)?

Really good point! I can only answer for Buidler EVM, where we always run with a listener attached to `step`, but we don't even read the values of the stack.

> Oh btw, the stack items were cloned outside of `runCode` after `opFns` returned, in the past. Is that not the case?

I can't find that. @s1na do you know if we still clone it?

**Comment by axic** [comment]
> Oh btw, the stack items were cloned outside of runCode after opFns returned, in the past. Is that not the case?

Maybe I remember wrongly and the rule was the first operation in the implementation would clone, but subsequent ones wouldn't.

That would be a sensible thing to do, but just removing every in-place operation probably is an overkill.

**Comment by ryanio** [comment]
I will close this PR since much of this code has been changed since the PR has been opened.

I think the team is now sensitive to how `bn.js` works, specifically that a `new BN()` uses the same underlying pointer and `clone()` should be used when explicitly needed.

**Pull request closed without merging**